osu.Game/Online/Chat/Channel.cs

```diff
@@ -52,11 +65,42 @@ public void AddNewMessages(params Message[] messages)
 
         private void purgeOldMessages()
         {
-            int messageCount = Messages.Count;
+            // never purge local echos
+            int messageCount = Messages.Count - pendingMessages.Count;
             if (messageCount > MAX_HISTORY)
                 Messages.RemoveRange(0, messageCount - MAX_HISTORY);
         }
 
+        /// <summary>
+        /// Replace or remove a message from the channel.
+        /// </summary>
+        /// <param name="echo">The local echo message (client-side).</param>
+        /// <param name="final">The response message, or null if the message became invalid.</param>
+        public void ReplaceMessage(LocalEchoMessage echo, Message final)
+        {
+            if (!pendingMessages.Remove(echo))
+                Trace.Fail("Attempted to remove echo that wasn't present");

```

Comment:
This should be an exception, not a trace. It is a possible route for a caller given that this is a public function.